File: ClientApp/src/app/Models/Job.ts
import {List} from "./List";
import {Project} from "./Project";
import {User, UserId} from "./User";

export interface JobStatus
{
  id: string;
  name: string;
  colour: string;
  user: User | null;
}

export interface NewStatus
{
  name: string;
  colour: string;
}

export interface JobPriority
{
  id: string;
  name: string;
  colour: string;
  user: User | null;
}

export interface NewPriority
{
  name: string;
  colour: string;
}

export interface JobType
{
  id: string;
  name: string;
  colour: string;
  user: User | null;
}

export interface NewType
{
  name: string;
  colour: string;
}

export interface JobEmployment
{
  jobId: string;
  userId: string;
}

export interface Job
{
  id?: string;
  title: string;
  description: string;
  created?: Date;
  deadline: Date;

  jobStatus: JobStatus;
  jobPriority: JobPriority;
  jobType: JobType;

  list?: List;
  project?: Project;
  employments?: JobEmployment[];
}

export interface NewJob
{
  job: Job;
  employments: User[];
  listId: string;
  projectId: string;
}

File: Models/Tenant/Project/Job/Job.cs
using System;
using System.Collections;
using System.Collections.Generic;
using Warehouse.Models.Joins;

namespace Warehouse.Models
{
    public class Job
    {
        public Guid Id { get; set; }
        public string Link { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public DateTime Created { get; set; }
        public DateTime Deadline { get; set; }
        public string AssociatedUrl { get; set; }
        public string Commit { get; set; }

        // many to one list
        public List List { get; set; }
        
        // many to one project
        public Project Project { get; set; }
        
        // many to many users
        public IList<JobEmployment> Employments { get; set; }
        
        // many to one fluff
        public JobStatus JobStatus { get; set; }
        public JobPriority JobPriority { get; set; }
        public JobType JobType { get; set; }
    }
    
    public class NewJob
    {
        public Job Job { get; set; }
        public IList<User> Employments { get; set; }
        public Guid ListId { get; set; }
        public Guid ProjectId { get; set; }
    }
    
    public class JobListViewModel
    {
        public Guid Id { get; set; }
        public string Link { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public DateTime Created { get; set; }
        public DateTime Deadline { get; set; }
        public string AssociatedUrl { get; set; }
        public string Commit { get; set; }
        
        // many to many users
        public IList<JobEmployment> Employments { get; set; }
        
        // many to one fluff
        public JobStatusViewModel JobStatus { get; set; }
        public JobPriorityViewModel JobPriority { get; set; }
        public JobTypeViewModel JobType { get; set; }
    }
}